relay-metrics/src/view.rs

```diff
@@ -0,0 +1,1061 @@
+/// of metric data slicing even into the buckets themselves.
+///
+/// ```txt
+///                    Full View
+///  /---------------------------------------------\
+/// [[C:1], [C:12], [D:0, 1, 2, 3, 5, 5], [S:42, 75]]
+///  \--------------------------/  \---------------/
+///            View 1                    View 2
+/// ```
+///
+/// Iterating over a [`BucketsView`] yields [`BucketView`] items,
+/// only the first and last elements may be partial buckets.
+///
+/// Using the above example, iterating over `View 1` yields the buckets:
+/// `[C:1], [C:12], [D:0, 1, 2, 3]`.
+pub struct BucketsView<'a> {
+    /// Source slice of buckets.
+    inner: &'a [Bucket],
+    /// Start index, slice index indicates bucket,
+    /// bucket index indicates offset in the selected bucket.
+    start: Index,
+    /// End index, slice index indicates exclusive end,
+    /// bucket index, indicates offset into the *next* bucket past the end.
+    end: Index,
+}
+
+impl<'a> BucketsView<'a> {
+    /// Creates a new buckets view containing all data from the slice.
+    pub fn new(buckets: &'a [Bucket]) -> Self {
+        Self {
+            inner: buckets,
+            start: Index {
+                slice: 0,
+                bucket: 0,
+            },
+            end: Index {
+                slice: buckets.len(),
+                bucket: 0,
+            },
+        }
+    }
+
+    /// Returns the amount of partial or full buckets in the view.
+    pub fn len(&self) -> usize {
+        let mut len = self.end.slice - self.start.slice;
+        if self.end.bucket != 0 {
+            len += 1;
+        }
+        len
+    }
+
+    /// Returns whether the view contains any buckets.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Iterater over all buckets in the view.
+    pub fn iter(&self) -> impl Iterator<Item = BucketView<'a>> {
+        BucketsViewIter::new(self.inner, self.start, self.end)
+    }
+
+    /// Iterator which slices the source view into segments with an approximate size of `size_in_bytes`.
+    pub fn by_size(&self, size_in_bytes: usize) -> impl Iterator<Item = BucketsView<'a>> {
+        BucketsViewBySizeIter::new(self.inner, self.start, self.end, size_in_bytes)
+    }
+}
+
+impl<'a> fmt::Debug for BucketsView<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let contents = self.iter().collect::<Vec<_>>();
+        f.debug_tuple("BucketsView").field(&contents).finish()
+    }
+}
+
+/// Iterator yielding all items contained in a [`BucketsView`].
+///
+/// First and/or last item may be partial buckets.
+struct BucketsViewIter<'a> {
+    /// Source slice of buckets.
+    inner: &'a [Bucket],
+    /// Current index.
+    current: Index,
+    /// End index.
+    end: Index,
+}
+
+impl<'a> BucketsViewIter<'a> {
+    /// Creates a new iterator.
+    ///
+    /// **Notes:** if `start` and `end` are not valid indices the iterator may panic.
```

Comment:
```suggestion
    /// # Panics
    /// 
    /// If `start` and `end` are not valid indices the iterator may panic.
```

Maybe we can add the proper `Panic` note on here, something like in the official rust docs used